egui/src/widgets/text_edit/builder.rs

```diff
@@ -248,6 +252,84 @@ impl<'t> TextEdit<'t> {
     }
 }
 
+/// Instructions on how the user input should be processed.
+#[non_exhaustive]
+pub enum Action {
+    /// Insert the string at the cursor.
+    ///
+    /// Note that inserting an empty string does nothing.
+    Insert(String),
+
+    /// Overwrite the entire buffer with the given string.
+    ///
+    /// This can be used for instance, to automatically fill in the
+    /// minimum or maximum value possible, if it is exceeded.
+    Overwrite(Cow<'static, str>),
+}
+
+/// The information required to validate text input.
+#[non_exhaustive]
+pub struct InputData<'a> {
+    /// The text that was input by the user.
+    pub input: String,
+
+    /// The current text in the text box buffer.
+    pub buffer: &'a str,
+
+    /// The current position of the start of the cursor.
+    ///
+    /// This represents where the data will be inserted.
+    pub cursor_start: usize,
```

Comment:
If a range of characters are selected than this is misleading. Perhaps remove it, or replace with `CursorRange` or `CCursorRange`.